clutch/src/lib.rs

```diff
@@ -0,0 +1,436 @@
+        self.apply_comm_aux(store, call_form, false)
+    }
+    fn call(&mut self, store: &mut Store<F>, rest: Ptr<F>) -> Result<Option<Ptr<F>>> {
+        self.apply_comm_aux(store, rest, false)
+    }
+    fn chain(&mut self, store: &mut Store<F>, rest: Ptr<F>) -> Result<Option<Ptr<F>>> {
+        let (result_expr, new_comm) = self
+            .apply_comm_aux(store, rest, true)?
+            .and_then(|cons| store.car_cdr(&cons).ok()) // unpack pair
+            .ok_or_else(|| {
+                anyhow!(
+                    "chained functional commitment output must be a pair (result, new-commitment)"
+                )
+            })?;
+
+        let new_secret = store
+            .secret(new_comm)
+            .ok_or_else(|| anyhow!("secret missing"))
+            .and_then(|hash| {
+                store
+                    .get_expr_hash(&hash)
+                    .ok_or_else(|| anyhow!("hash missing"))
+            })
+            .map(|hash| *hash.value())?;
+
+        let (_, new_fun) = store
+            .open(new_comm)
+            .ok_or_else(|| anyhow!("opening missing"))?;
+
+        let new_commitment = Commitment::from_comm(store, &new_comm);
+
+        let expr = LurkPtr::from_ptr(store, &new_fun);
+
+        let new_function = CommittedExpression::<F> {
+            expr,
+            secret: Some(new_secret),
+            commitment: Some(new_commitment),
+        };
+
+        self.expression_map.set(new_commitment, &new_function)?;
+
+        let interned_commitment = store.intern_maybe_opaque_comm(new_commitment.comm);
+        let mut handle = io::stdout().lock();
+        interned_commitment.fmt(store, &mut handle)?;
+        println!();
+
+        Ok(Some(result_expr))
+    }
+    fn apply_comm_aux(
+        &mut self,
+        store: &mut Store<F>,
+        rest: Ptr<F>,
+        chain: bool,
+    ) -> Result<Option<Ptr<F>>> {
+        let args = store.cdr(&rest)?;
+        if let (commitment, Some(e)) = self.open_aux(store, rest)? {
+            let call = store.cons(e, args);
+            let (arg, _) = store.car_cdr(&args)?;
+
+            let (result, _iterations, cont, _) = self
+                .repl_state
+                .eval_expr(call, store)
+                .with_context(|| "Evaluating call")?;
+
+            let (output, new_commitment) = if chain {
+                let (output, new_comm) = store.car_cdr(&result)?;
+                (output, Some(Commitment::from_comm(store, &new_comm)))
+            } else {
+                (result, None)
+            };
+
+            let claim = Claim::Opening::<F>(Opening {
+                input: arg.fmt_to_string(store),
+                output: output.fmt_to_string(store),
+                status: Status::from(cont),
+                commitment,
+                new_commitment,
+            });
+
+            self.last_claim = Some(claim);
+
+            Ok(Some(result))
+        } else {
+            Ok(None)
+        }
+    }
```

Comment:
Nit
```suggestion
    fn apply_comm_aux(
        &mut self,
        store: &mut Store<F>,
        rest: Ptr<F>,
        chain: bool,
    ) -> Result<Option<Ptr<F>>> {
        let args = store.cdr(&rest)?;
        let (commitment, Some(e)) = self.open_aux(store, rest)? else {
            return Ok(None);
        };

        let call = store.cons(e, args);
        let (arg, _) = store.car_cdr(&args)?;

        let (result, _iterations, cont, _) = self
            .repl_state
            .eval_expr(call, store)
            .with_context(|| "Evaluating call")?;

        let (output, new_commitment) = if chain {
            let (output, new_comm) = store.car_cdr(&result)?;
            (output, Some(Commitment::from_comm(store, &new_comm)))
        } else {
            (result, None)
        };

        let claim = Claim::Opening::<F>(Opening {
            input: arg.fmt_to_string(store),
            output: output.fmt_to_string(store),
            status: Status::from(cont),
            commitment,
            new_commitment,
        });

        self.last_claim = Some(claim);

        Ok(Some(result))
    }
```

Reply:
Thanks, I missed the existence of this helpfulf construct.